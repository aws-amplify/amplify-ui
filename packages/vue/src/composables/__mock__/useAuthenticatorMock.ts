import { AuthenticatorServiceFacade } from '@aws-amplify/ui';

export const baseMockServiceFacade: AuthenticatorServiceFacade = {
  authStatus: 'authenticated',
  codeDeliveryDetails: {} as AuthenticatorServiceFacade['codeDeliveryDetails'],
  error: undefined as unknown as AuthenticatorServiceFacade['error'],
  hasValidationErrors: false,
  isPending: false,
  route: 'idle',
  socialProviders: [],
  unverifiedContactMethods: { email: 'test#example.com' },
  user: {} as AuthenticatorServiceFacade['user'],
  validationErrors:
    {} as unknown as AuthenticatorServiceFacade['validationErrors'],
  totpSecretCode: null,
  initializeMachine: jest.fn(),
  resendCode: jest.fn(),
  signOut: jest.fn(),
  submitForm: jest.fn(),
  updateForm: jest.fn(),
  updateBlur: jest.fn(),
  toFederatedSignIn: jest.fn(),
  toResetPassword: jest.fn(),
  toSignIn: jest.fn(),
  toSignUp: jest.fn(),
  skipVerification: jest.fn(),
};
