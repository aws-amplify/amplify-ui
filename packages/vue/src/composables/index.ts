export * from './useAuth';
// export * from "./useAuth";
