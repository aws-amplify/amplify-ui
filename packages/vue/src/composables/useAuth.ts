import { createSharedComposable } from '@vueuse/core';
import { ref, reactive, Ref, watchEffect, onScopeDispose } from 'vue';
import { useActor } from '@xstate/vue';
import { interpret } from 'xstate';

import { Auth } from 'aws-amplify';
import {
  AuthInterpreter,
  AuthStatus,
  createAuthenticatorMachine,
  defaultAuthHubHandler,
  getServiceFacade,
  listenToAuthHub,
} from '@aws-amplify/ui';

import { UseAuth } from '../types';

export const useAuth = createSharedComposable((): UseAuth => {
  const machine = createAuthenticatorMachine();
  const service: AuthInterpreter = interpret(machine).start();
  const authStatus: Ref<AuthStatus> = ref('configuring');

  const { state, send } = useActor(service);

  const onSignIn = () => {
    authStatus.value = 'authenticated';
  };

  const onSignOut = () => {
    authStatus.value = 'unauthenticated';
  };

  const unsubscribeHub = listenToAuthHub(service, async (data, service) => {
    await defaultAuthHubHandler(data, service, { onSignIn, onSignOut });
  });

  Auth.currentAuthenticatedUser()
    .then(() => {
      authStatus.value = 'authenticated';
    })
    .catch(() => {
      authStatus.value = 'unauthenticated';
    });

  onScopeDispose(() => {
    unsubscribeHub();
  });

  return { authStatus, service, send, state };
});

export const useAuthenticator = createSharedComposable(() => {
  const { authStatus, state, send } = useAuth();

  // TODO(BREAKING): remove the cast to any
  const useAuthenticatorValue = reactive({}) as any;

  /*
   * Note that watchEffect runs immediately, so `useAuthenticatorValue` is
   * guaranteed to have facade values by the time `useAuthenticator` returns.
   *
   * https://vuejs.org/api/reactivity-core.html#watcheffect
   */
  watchEffect(() => {
    const facade = getServiceFacade({ send, state: state.value });

    /*
     * TODO(BREAKING): consider using a plain object with `refs` instead of
     * one `reactive` object to prevent iterating manually over its keys.
     */
    for (const key of Object.keys(facade)) {
      //@ts-ignore
      useAuthenticatorValue[key] = facade[key];
    }
    useAuthenticatorValue.authStatus = authStatus.value;
    useAuthenticatorValue.send = send;
    useAuthenticatorValue.state = state;
  });

  return useAuthenticatorValue;
});
