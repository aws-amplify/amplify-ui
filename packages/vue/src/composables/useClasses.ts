import { inject } from "vue";

import { classNames } from "../defaults/DefaultClasses";

export function useClasses(val: string, category?, defaultClasses?): any {
  let classes: any;
  if (!category) {
    category = inject("category");
    defaultClasses = inject("defaultClasses");
  }
  classes = {};
  if (defaultClasses?.value) {
    // grab default classes
    classes = {
      ...classNames,
    };
  }
  return classes[`${category?.value}.${val}`];
}
