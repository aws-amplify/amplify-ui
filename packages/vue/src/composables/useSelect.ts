import { ref } from 'vue';

// state

const active = ref(false);

export default { active };
