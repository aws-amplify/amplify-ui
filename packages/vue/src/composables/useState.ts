import { ref, readonly } from 'vue';

// State
const currentPage = ref('SIGNIN');

//mutatation

const setPage = (page) => (currentPage.value = page);

export default { currentPage: readonly(currentPage), setPage };
