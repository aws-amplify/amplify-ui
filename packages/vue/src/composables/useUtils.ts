import {
  AuthEventData,
  LoginMechanism,
  LoginMechanismArray,
} from '@aws-amplify/ui';

/**
 * If 'username' is the only login mechanism, then we ask for a user's
 * email and phone number during sign up as well.
 */
export const useAliases = (
  login_mechanisms: LoginMechanism[]
): LoginMechanism[] =>
  login_mechanisms?.length === 1 && login_mechanisms[0] === 'username'
    ? [...LoginMechanismArray]
    : login_mechanisms;

export const facade = {
  error: '',
  hasValidationErrors: false,
  isPending: false,
  route: '',
  user: '',
  validationErrors: {
    val: '',
  },
  codeDeliveryDetails: {
    val: '',
  },
  change: (data?: AuthEventData) => null,
  updateForm: (data?: AuthEventData) => null,
  updateBlur: (data?: AuthEventData) => null,
  resend: (data?: AuthEventData) => null,
  resendCode: (data?: AuthEventData) => null,
  signOut: (data?: AuthEventData) => null,
  federatedSignIn: (data?: AuthEventData) => null,
  toFederatedSignIn: (data?: AuthEventData) => null,
  resetPassword: (data?: AuthEventData) => null,
  toResetPassword: (data?: AuthEventData) => null,
  signIn: (data?: AuthEventData) => null,
  toSignIn: (data?: AuthEventData) => null,
  signUp: (data?: AuthEventData) => null,
  toSignUp: (data?: AuthEventData) => null,
  skip: (data?: AuthEventData) => null,
  skipVerification: (data?: AuthEventData) => null,
  submit: (data?: AuthEventData) => null,
  submitForm: (data?: AuthEventData) => null,
};
