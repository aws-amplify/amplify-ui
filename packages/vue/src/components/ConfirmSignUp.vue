<template>
  <slot name="confirmSignUpSlotI">
    <Form @submit.prevent="onConfirmSignUpSubmit">
      <Heading>
        {{ confirmSignUpHeading }}
      </Heading>
      <FieldSet>
        <SignUpUsernameControl
          :userName="state?.context?.user?.user?.username"
          :disabled="true"
        />
        <Label data-amplify-password>
          <Text>{{ confirmationCodeText }}</Text>
          <Input name="code" required type="number"></Input>
          <Box>
            <Text> {{ lostYourCodeText }}</Text>
            <Button type="button" @click.prevent="onLostCodeClicked">
              {{ resendCodeText }}
            </Button>
          </Box>
        </Label>
      </FieldSet>

      <Footer>
        <Button type="button" @click.prevent="onBackToSignInClicked">
          {{ backSignInText }}</Button
        >
        <Spacer />
        <Button>{{ confirmText }}</Button>
      </Footer>
    </Form>
  </slot>
</template>

<script lang="ts">
import { defineComponent, computed } from "vue";
import Heading from "./primitives/Heading.vue";
import FieldSet from "./primitives/FieldSet.vue";
import SignUpUsernameControl from "./SignUpUsernameControl.vue";
import Label from "./primitives/Label.vue";

import Spacer from "./primitives/Spacer.vue";
import Button from "./primitives/Button.vue";
import Footer from "./primitives/Footer.vue";
import Text from "./primitives/Text.vue";
import Input from "./primitives/Input.vue";
import Form from "./primitives/Form.vue";
import Box from "./primitives/Box.vue";

import {
  CONFIRM_SIGNUP_HEADING,
  CONFIRMATION_CODE_TEXT,
  LOST_YOUR_CODE_TEXT,
  RESEND_CODE_TEXT,
  BACK_SIGN_IN_TEXT,
  CONFIRM_TEXT
} from "../defaults/DefaultTexts";
import { useAuth } from "../composables/useAuth";

export default defineComponent({
  components: {
    Box,
    Heading,
    FieldSet,
    Form,
    SignUpUsernameControl,
    Label,
    Spacer,
    Button,
    Footer,
    Text,
    Input
  },
  inheritAttrs: false,
  setup(
    _,
    {
      emit,
      attrs
    }: { emit: (st, e?) => unknown; attrs: Record<string, unknown> }
  ) {
    const { state, send } = useAuth();

    //computed properties

    const confirmSignUpHeading = computed(() => CONFIRM_SIGNUP_HEADING);
    const confirmationCodeText = computed(() => CONFIRMATION_CODE_TEXT);
    const lostYourCodeText = computed(() => LOST_YOUR_CODE_TEXT);
    const resendCodeText = computed(() => RESEND_CODE_TEXT);
    const backSignInText = computed(() => BACK_SIGN_IN_TEXT);
    const confirmText = computed(() => CONFIRM_TEXT);

    // Methods
    const onConfirmSignUpSubmit = (e): void => {
      if (attrs?.onConfirmSignUpSubmit) {
        emit("confirmSignUpSubmit");
      } else {
        submit(e);
      }
    };

    const submit = (e): void => {
      const formData = new FormData(e.target);
      send({
        type: "SUBMIT",
        //@ts-ignore
        data: {
          //@ts-ignore
          ...Object.fromEntries(formData),
          username: state?.value.context?.user?.user?.username
        }
      });
    };

    const onLostCodeClicked = (): void => {
      // do something
      if (attrs?.onLostCodeClicked) {
        emit("lostCodeClicked");
      } else {
        send({
          type: "RESEND",
          // @ts-ignore
          data: { username: state?.value?.context?.user?.user?.username }
        });
      }
    };

    const onBackToSignInClicked = (): void => {
      if (attrs?.onBackToSignInClicked) {
        emit("backToSignInClicked");
      } else {
        send({
          type: "SIGN_IN"
        });
      }
    };

    return {
      onConfirmSignUpSubmit,
      onBackToSignInClicked,
      submit,
      confirmSignUpHeading,
      confirmationCodeText,
      lostYourCodeText,
      resendCodeText,
      backSignInText,
      confirmText,
      onLostCodeClicked,
      state,
      send
    };
  }
});
</script>

<style scoped></style>
