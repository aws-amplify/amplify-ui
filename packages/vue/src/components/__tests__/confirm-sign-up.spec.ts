import { reactive, Ref, ref } from 'vue';
import { fireEvent, render, screen } from '@testing-library/vue';

import * as UIModule from '@aws-amplify/ui';
import {
  AuthenticatorServiceFacade,
  AuthInterpreter,
  AuthMachineState,
} from '@aws-amplify/ui';

import { components } from '../../../global-spec';
import * as UseAuthComposables from '../../composables/useAuth';
import { baseMockServiceFacade } from '../../composables/__mock__/useAuthenticatorMock';
import ConfirmSignUp from '../confirm-sign-up.vue';

jest.spyOn(UseAuthComposables, 'useAuth').mockReturnValue({
  authStatus: ref('unauthenticated'),
  send: jest.fn(),
  service: undefined as unknown as AuthInterpreter,
  state: ref(undefined) as unknown as Ref<AuthMachineState>,
});

jest.spyOn(UseAuthComposables, 'useAuth').mockReturnValue({
  authStatus: ref('unauthenticated'),
  send: jest.fn(),
  service: undefined as unknown as AuthInterpreter,
  state: ref(undefined) as unknown as Ref<AuthMachineState>,
});

const updateFormSpy = jest.fn();
const submitFormSpy = jest.fn();
const resendCodeSpy = jest.fn();

const mockServiceFacade: AuthenticatorServiceFacade = {
  ...baseMockServiceFacade,
  route: 'confirmSignUp',
  updateForm: updateFormSpy,
  submitForm: submitFormSpy,
  resendCode: resendCodeSpy,
};

const useAuthenticatorSpy = jest
  .spyOn(UseAuthComposables, 'useAuthenticator')
  .mockReturnValue(reactive(mockServiceFacade));

jest.spyOn(UIModule, 'getActorContext').mockReturnValue({
  country_code: '+1',
});

jest.spyOn(UIModule, 'getSortedFormFields').mockReturnValue([
  [
    'confirmation_code',
    {
      label: 'Confirmation Code',
      placeholder: 'Enter your code',
      type: 'number',
    },
  ],
]);

const codeInputParams = { name: 'confirmation_code', value: '123456' };

describe('ConfirmSignUp', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', () => {
    // mock random value so that snapshots are consistent
    const mathRandomSpy = jest.spyOn(Math, 'random').mockReturnValue(0.1);

    const { container } = render(ConfirmSignUp, { global: { components } });
    expect(container).toMatchSnapshot();

    mathRandomSpy.mockRestore();
  });

  it('sends change event on form input', async () => {
    render(ConfirmSignUp, { global: { components } });

    const codeField = await screen.findByLabelText('Confirmation Code');

    await fireEvent.input(codeField, { target: codeInputParams });
    expect(updateFormSpy).toHaveBeenCalledWith(codeInputParams);
  });

  it('sends submit event on form submit', async () => {
    render(ConfirmSignUp, { global: { components } });

    const codeField = await screen.findByLabelText('Confirmation Code');

    await fireEvent.input(codeField, { target: codeInputParams });
    expect(updateFormSpy).toHaveBeenCalledWith(codeInputParams);

    const submitButton = await screen.findByRole('button', {
      name: 'Confirm',
    });
    await fireEvent.click(submitButton);
    expect(submitFormSpy).toHaveBeenCalledTimes(1);
  });

  it('displays error if it is present', async () => {
    useAuthenticatorSpy.mockReturnValueOnce(
      reactive({
        ...mockServiceFacade,
        error: 'mockError',
      })
    );
    render(ConfirmSignUp, { global: { components } });

    expect(await screen.findByText('mockError')).toBeInTheDocument();
  });

  it('handles resend code button as expected', async () => {
    render(ConfirmSignUp, { global: { components } });

    const resendCodeButton = await screen.findByRole('button', {
      name: 'Resend Code',
    });

    await fireEvent.click(resendCodeButton);

    expect(resendCodeSpy).toHaveBeenCalledTimes(1);
  });

  it('disables the submit button if if sign up is pending', async () => {
    useAuthenticatorSpy.mockReturnValue(
      reactive({ ...mockServiceFacade, isPending: true })
    );
    render(ConfirmSignUp, { global: { components } });

    const submitButton = await screen.findByRole('button', {
      name: 'Confirm',
    });
    expect(submitButton).toBeDisabled();
  });
});
