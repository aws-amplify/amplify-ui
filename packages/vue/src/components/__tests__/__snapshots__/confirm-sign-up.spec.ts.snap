// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmSignUp renders as expected 1`] = `
<div>
  
  <div>
    
    
    <form
      data-amplify-form=""
    >
      
      <div
        class="amplify-flex amplify-authenticator__column"
      >
        
        
        <h3
          class="amplify-heading amplify-authenticator__heading amplify-heading--3"
          data-amplify-heading=""
        >
          We Texted You
        </h3>
        
        
        <span
          class="amplify-authenticator__subtitle"
          data-amplify-text=""
        >
          
          Your code is on the way. To log in, enter the code we sent you. It may take a minute to arrive.
          
        </span>
        
        
        <fieldset
          class="amplify-flex amplify-authenticator__column"
          data-amplify-fieldset=""
        >
          
          
          
          <!-- password input -->
          
          <!-- textfield input -->
          <div
            class="amplify-flex amplify-field amplify-textfield amplify-phonenumberfield amplify-authenticator__column"
          >
            
            <label
              class="amplify-label"
              data-amplify-label=""
              for="amplify-field-99999"
            >
              
              Confirmation Code
              
            </label>
            <div
              class="amplify-flex amplify-field-group"
            >
              
              <div
                class="amplify-field-group__outer-start"
              >
                
                
              </div>
              <div
                class="amplify-field-group__field-wrapper"
              >
                
                <!--Phone input-->
                <input
                  aria-invalid="false"
                  autocomplete=""
                  class="amplify-input amplify-field-group__control"
                  data-amplify-input=""
                  id="amplify-field-99999"
                  name="confirmation_code"
                  placeholder="Enter your code"
                  required=""
                  type="number"
                />
                
              </div>
              
            </div>
            
          </div>
          
          <!-- Validation error, if any -->
          <!--v-if-->
          
          
          
        </fieldset>
        
        
        <footer
          class="amplify-flex amplify-authenticator__column"
        >
          
          <!--v-if-->
          
          <button
            class="amplify-button amplify-button--primary amplify-field-group__control amplify-authenticator__font"
            data-amplify-button=""
            data-disabled="false"
            data-fullwidth="false"
            data-loading="false"
            data-variation="primary"
            fullwidth="false"
            type="submit"
          >
            
            Confirm
            
          </button>
          
          
          <button
            class="amplify-button amplify-button--default amplify-field-group__control amplify-authenticator__font"
            data-amplify-button=""
            data-disabled="false"
            data-fullwidth="false"
            data-loading="false"
            data-variation="default"
            fullwidth="false"
            style="font-weight: normal;"
            type="button"
          >
            
            Resend Code
            
          </button>
          
          
          
          
        </footer>
        
        
      </div>
      
    </form>
    
    
  </div>
  
</div>
`;
